In [3]:
import json

with open("../data/train_ner_data.json", "r", encoding="utf-8") as f:
    train_ner_data = json.load(f)

with open("../data/val_ner_data.json", "r", encoding="utf-8") as f:
    val_ner_data = json.load(f)

with open("../data/test_ner_data.json", "r", encoding="utf-8") as f:
    test_ner_data = json.load(f)

print("Train:", len(train_ner_data))
print("Validation:", len(val_ner_data))
print("Test:", len(test_ner_data))

Train: 140
Validation: 30
Test: 30


In [4]:



all_labels = set()
for data in train_ner_data:
    all_labels.update(data["labels"])

all_labels = sorted(list(all_labels))

print("Number of unique labels:", len(all_labels))
print("Unique labels:", all_labels)

Number of unique labels: 82
Unique labels: ['B-Activity', 'B-Administration', 'B-Age', 'B-Area', 'B-Biological_attribute', 'B-Biological_structure', 'B-Clinical_event', 'B-Color', 'B-Coreference', 'B-Date', 'B-Detailed_description', 'B-Diagnostic_procedure', 'B-Disease_disorder', 'B-Distance', 'B-Dosage', 'B-Duration', 'B-Family_history', 'B-Frequency', 'B-Height', 'B-History', 'B-Lab_value', 'B-Mass', 'B-Medication', 'B-Nonbiological_location', 'B-Occupation', 'B-Other_entity', 'B-Other_event', 'B-Outcome', 'B-Personal_background', 'B-Qualitative_concept', 'B-Quantitative_concept', 'B-Severity', 'B-Sex', 'B-Shape', 'B-Sign_symptom', 'B-Subject', 'B-Texture', 'B-Therapeutic_procedure', 'B-Time', 'B-Volume', 'B-Weight', 'I-Activity', 'I-Administration', 'I-Age', 'I-Area', 'I-Biological_attribute', 'I-Biological_structure', 'I-Clinical_event', 'I-Color', 'I-Coreference', 'I-Date', 'I-Detailed_description', 'I-Diagnostic_procedure', 'I-Disease_disorder', 'I-Distance', 'I-Dosage', 'I-Durat

In [5]:
label2id = {
    label : i
    for i, label in enumerate(all_labels)
}

label2id

{'B-Activity': 0,
 'B-Administration': 1,
 'B-Age': 2,
 'B-Area': 3,
 'B-Biological_attribute': 4,
 'B-Biological_structure': 5,
 'B-Clinical_event': 6,
 'B-Color': 7,
 'B-Coreference': 8,
 'B-Date': 9,
 'B-Detailed_description': 10,
 'B-Diagnostic_procedure': 11,
 'B-Disease_disorder': 12,
 'B-Distance': 13,
 'B-Dosage': 14,
 'B-Duration': 15,
 'B-Family_history': 16,
 'B-Frequency': 17,
 'B-Height': 18,
 'B-History': 19,
 'B-Lab_value': 20,
 'B-Mass': 21,
 'B-Medication': 22,
 'B-Nonbiological_location': 23,
 'B-Occupation': 24,
 'B-Other_entity': 25,
 'B-Other_event': 26,
 'B-Outcome': 27,
 'B-Personal_background': 28,
 'B-Qualitative_concept': 29,
 'B-Quantitative_concept': 30,
 'B-Severity': 31,
 'B-Sex': 32,
 'B-Shape': 33,
 'B-Sign_symptom': 34,
 'B-Subject': 35,
 'B-Texture': 36,
 'B-Therapeutic_procedure': 37,
 'B-Time': 38,
 'B-Volume': 39,
 'B-Weight': 40,
 'I-Activity': 41,
 'I-Administration': 42,
 'I-Age': 43,
 'I-Area': 44,
 'I-Biological_attribute': 45,
 'I-Biological_s

In [6]:
id2label = {
    i : label
    for label, i in label2id.items()
}

id2label

{0: 'B-Activity',
 1: 'B-Administration',
 2: 'B-Age',
 3: 'B-Area',
 4: 'B-Biological_attribute',
 5: 'B-Biological_structure',
 6: 'B-Clinical_event',
 7: 'B-Color',
 8: 'B-Coreference',
 9: 'B-Date',
 10: 'B-Detailed_description',
 11: 'B-Diagnostic_procedure',
 12: 'B-Disease_disorder',
 13: 'B-Distance',
 14: 'B-Dosage',
 15: 'B-Duration',
 16: 'B-Family_history',
 17: 'B-Frequency',
 18: 'B-Height',
 19: 'B-History',
 20: 'B-Lab_value',
 21: 'B-Mass',
 22: 'B-Medication',
 23: 'B-Nonbiological_location',
 24: 'B-Occupation',
 25: 'B-Other_entity',
 26: 'B-Other_event',
 27: 'B-Outcome',
 28: 'B-Personal_background',
 29: 'B-Qualitative_concept',
 30: 'B-Quantitative_concept',
 31: 'B-Severity',
 32: 'B-Sex',
 33: 'B-Shape',
 34: 'B-Sign_symptom',
 35: 'B-Subject',
 36: 'B-Texture',
 37: 'B-Therapeutic_procedure',
 38: 'B-Time',
 39: 'B-Volume',
 40: 'B-Weight',
 41: 'I-Activity',
 42: 'I-Administration',
 43: 'I-Age',
 44: 'I-Area',
 45: 'I-Biological_attribute',
 46: 'I-Biologic

In [7]:
import sys

print(sys.executable)
print(sys.version)

c:\Users\gangu\healthcare-clinical-nlp\.venv\Scripts\python.exe
3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]


In [8]:
import transformers

print(transformers.__version__)
print(transformers.__file__)

5.16.1
c:\Users\gangu\healthcare-clinical-nlp\.venv\Lib\site-packages\transformers\__init__.py


In [9]:
from transformers import AutoTokenizer

In [10]:
import sys
print(sys.executable)

from transformers import AutoTokenizer
print("AutoTokenizer imported successfully")

c:\Users\gangu\healthcare-clinical-nlp\.venv\Scripts\python.exe
AutoTokenizer imported successfully


In [11]:
model_name = "emilyalsentzer/Bio_ClinicalBERT"

tokenizer = AutoTokenizer.from_pretrained(model_name)


In [12]:
example_tokens = [
    "A",
    "31",
    "-",
    "year",
    "-",
    "old",
    "man",
    "developed",
    "diabetes",
    "insipidus"
]

encoded = tokenizer(
    example_tokens,
    is_split_into_words=True
)

bert_tokens = tokenizer.convert_ids_to_tokens(
    encoded["input_ids"]
)

print(bert_tokens)

['[CLS]', 'a', '31', '-', 'year', '-', 'old', 'man', 'developed', 'diabetes', 'ins', '##ip', '##id', '##us', '[SEP]']


In [13]:
word_ids = encoded.word_ids()

print(word_ids)

[None, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 9, 9, 9, None]


In [14]:
def tokenize_and_align_labels(tokens, labels):
    tokenized_inputs = tokenizer(
        tokens,
        is_split_into_words=True,
        padding="max_length",
        truncation=True
    )

    word_ids = tokenized_inputs.word_ids()

    aligned_labels = []
    previous_word_id = None

    for word_id in word_ids:

        if word_id is None:
            aligned_labels.append(-100)
        elif word_id != previous_word_id:
            label = labels[word_id]
            aligned_labels.append(label2id[label])
        else:
            aligned_labels.append(-100)

        previous_word_id = word_id
    tokenized_inputs["labels"] = aligned_labels


    return tokenized_inputs

In [15]:
sample = train_ner_data[0]

result = tokenize_and_align_labels(
    sample["tokens"],
    sample["labels"]
)

print("Input IDs:", len(result["input_ids"]))
print("Labels:", len(result["labels"]))

Input IDs: 459
Labels: 459


In [16]:
bert_tokens = tokenizer.convert_ids_to_tokens(result["input_ids"])

for token, label_id in zip(
    bert_tokens[:40],
    result["labels"][:40]
):
    if label_id == -100:
        label_name = "Ignored"
    else:
        label_name = id2label[label_id]
    print(f"{token:20} {label_name}")

[CLS]                Ignored
case                 O
:                    O
a                    O
28                   B-Age
-                    I-Age
year                 I-Age
-                    I-Age
old                  I-Age
previously           B-History
healthy              I-History
man                  B-Sex
presented            B-Clinical_event
with                 O
a                    O
6                    B-Duration
-                    I-Duration
week                 I-Duration
history              O
of                   O
p                    B-Sign_symptom
##al                 Ignored
##pit                Ignored
##ations             Ignored
.                    O
the                  O
symptoms             B-Coreference
occurred             O
during               O
rest                 B-Clinical_event
,                    O
2                    B-Frequency
–                    Ignored
3                    Ignored
times                I-Frequency
per              

In [17]:
tokenized_train_data = [
    tokenize_and_align_labels(
        item["tokens"],
        item["labels"]
    )
    for item in train_ner_data
]

tokenized_val_data = [
    tokenize_and_align_labels(
        item["tokens"],
        item["labels"]
    )
    for item in val_ner_data
]

tokenized_test_data = [
    tokenize_and_align_labels(
        item["tokens"],
        item["labels"]
    )
    for item in test_ner_data
]

In [18]:
lengths = []

for item in train_ner_data:
    encoded = tokenizer(
        item["tokens"],
        is_split_into_words = True,
        trucation = False
    )

    lengths.append(len(encoded["input_ids"]))

print("Maximum length:", max(lengths))
print("Minimum length:", min(lengths))
print("Average length:", sum(lengths)/ len(lengths))

print("Documents over 512 tokens:", sum(length > 512 for length in lengths))

Maximum length: 1567
Minimum length: 229
Average length: 720.75
Documents over 512 tokens: 107


In [19]:
def tokenize_and_align_labels_with_chunks(tokens, labels):
    tokenized_inputs = tokenizer(
        tokens,
        is_split_into_words=True,
        truncation=True,
        max_length=512,
        stride=128,
        return_overflowing_tokens=True
    )

    all_labels = []

    for chunk_index in range(len(tokenized_inputs["input_ids"])):

        word_ids = tokenized_inputs.word_ids(
            batch_index=chunk_index
        )

        aligned_labels = []
        previous_word_id = None

        for word_id in word_ids:

            if word_id is None:
                aligned_labels.append(-100)

            elif word_id != previous_word_id:
                label = labels[word_id]
                aligned_labels.append(
                    label2id[label]
                )

            else:
                aligned_labels.append(-100)

            previous_word_id = word_id

        all_labels.append(aligned_labels)

    tokenized_inputs["labels"] = all_labels

    return tokenized_inputs

In [20]:
long_sample = max(
    train_ner_data,
    key=lambda x: len(
        tokenizer(
            x["tokens"],
            is_split_into_words=True,
            truncation=False
        )["input_ids"]
    )
)

chunked_result = tokenize_and_align_labels_with_chunks(
    long_sample["tokens"],
    long_sample["labels"]
)

In [21]:
print(
    "Number of chunks:",
    len(chunked_result["input_ids"])
)

Number of chunks: 4


In [22]:
for i in range(
    len(chunked_result["input_ids"])
):

    input_length = len(
        chunked_result["input_ids"][i]
    )

    label_length = len(
        chunked_result["labels"][i]
    )

    print(
        f"Chunk {i}: "
        f"Input IDs = {input_length}, "
        f"Labels = {label_length}"
    )

Chunk 0: Input IDs = 512, Labels = 512
Chunk 1: Input IDs = 512, Labels = 512
Chunk 2: Input IDs = 512, Labels = 512
Chunk 3: Input IDs = 421, Labels = 421


In [23]:
def prepare_chunked_dataset(dataset):
    chunked_dataset = []

    for item in dataset:
        result = tokenize_and_align_labels_with_chunks(
            item["tokens"],
            item["labels"]
        )

        for i in range(len(result["input_ids"])):
            chunked_dataset.append({
                "input_ids": result["input_ids"][i],
                "attention_mask": result["attention_mask"][i],
                "labels": result["labels"][i]
            })

    return chunked_dataset

In [24]:
train_chunked = prepare_chunked_dataset(train_ner_data)
val_chunked = prepare_chunked_dataset(val_ner_data)
test_chunked = prepare_chunked_dataset(test_ner_data)

print("Train chunks:", len(train_chunked))
print("Validation chunks:", len(val_chunked))
print("Test chunks:", len(test_chunked))

Train chunks: 286
Validation chunks: 64
Test chunks: 59


In [25]:
from datasets import Dataset

In [26]:
import sys
print(sys.executable)

c:\Users\gangu\healthcare-clinical-nlp\.venv\Scripts\python.exe


In [27]:
%pip install datasets

Note: you may need to restart the kernel to use updated packages.


In [28]:
train_dataset = Dataset.from_list(train_chunked)
val_dataset = Dataset.from_list(val_chunked)
test_dataset = Dataset.from_list(test_chunked)

print(train_dataset)
print(val_dataset)
print(test_dataset)

Dataset({
    features: ['input_ids', 'attention_mask', 'labels'],
    num_rows: 286
})
Dataset({
    features: ['input_ids', 'attention_mask', 'labels'],
    num_rows: 64
})
Dataset({
    features: ['input_ids', 'attention_mask', 'labels'],
    num_rows: 59
})


In [29]:
from transformers import AutoModelForTokenClassification

model = AutoModelForTokenClassification.from_pretrained(
    model_name,
    num_labels=len(all_labels),
    id2label=id2label,
    label2id=label2id
)

Loading weights: 100%|██████████| 197/197 [00:00<00:00, 21714.44it/s]
[transformers] BertForTokenClassification LOAD REPORT from: emilyalsentzer/Bio_ClinicalBERT
Key                                        | Status     | 
-------------------------------------------+------------+-
cls.predictions.bias                       | UNEXPECTED | 
cls.seq_relationship.bias                  | UNEXPECTED | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED | 
bert.pooler.dense.weight                   | UNEXPECTED | 
cls.predictions.decoder.weight             | UNEXPECTED | 
cls.seq_relationship.weight                | UNEXPECTED | 
bert.pooler.dense.bias                     | UNEXPECTED | 
cls.predictions.transform.dense.weight     | UNEXPECTED | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED | 
cls.predictions.transform.dense.bias       | UNEXPECTED | 
classifier.weight                          | MISSING    | 
classifier.bias                            | MISSING    | 

Notes:
- UN

In [30]:
print(type(model))
print("Number of labels:", model.config.num_labels)

<class 'transformers.models.bert.modeling_bert.BertForTokenClassification'>
Number of labels: 82


In [31]:
print(model.classifier)

Linear(in_features=768, out_features=82, bias=True)


In [32]:
from transformers import DataCollatorForTokenClassification

data_collator = DataCollatorForTokenClassification(
    tokenizer=tokenizer
)

In [33]:
sample_batch = [
    train_dataset[0],
    train_dataset[-1]
]

print("Before collator:")

for i, sample in enumerate(sample_batch):
    print(f"Example {i}:", len(sample["input_ids"]))

Before collator:
Example 0: 459
Example 1: 225


In [34]:
batch = data_collator(sample_batch)

print("\nAfter collator:")
print("Input IDs shape:", batch["input_ids"].shape)
print("Attention mask shape:", batch["attention_mask"].shape)
print("Labels shape:", batch["labels"].shape)


After collator:
Input IDs shape: torch.Size([2, 459])
Attention mask shape: torch.Size([2, 459])
Labels shape: torch.Size([2, 459])


In [35]:
print(
    "Padding positions in example 0:",
    (batch["attention_mask"][0] == 0).sum().item()
)

print(
    "Padding positions in example 1:",
    (batch["attention_mask"][1] == 0).sum().item()
)

Padding positions in example 0: 0
Padding positions in example 1: 234


In [36]:
print(
    "Number of -100 labels in example 1:",
    (batch["labels"][1] == -100).sum().item()
)

Number of -100 labels in example 1: 290


## Evaluation Metrics

In [37]:
%pip install seqeval

Note: you may need to restart the kernel to use updated packages.


In [38]:
from seqeval.metrics import precision_score, recall_score, f1_score

In [39]:
def compute_metrics(eval_pred):

    predictions, labels = eval_pred

    predictions = np.argmax(
        predictions,
        axis=2
    )

    true_predictions = []
    true_labels = []

    for prediction, label in zip(predictions, labels):

        current_predictions = []
        current_labels = []

        for predicted_id, label_id in zip(prediction, label):

            if label_id != -100:

                current_predictions.append(
                    id2label[int(predicted_id)]
                )

                current_labels.append(
                    id2label[int(label_id)]
                )

        true_predictions.append(current_predictions)
        true_labels.append(current_labels)

    precision = precision_score(
        true_labels,
        true_predictions
    )

    recall = recall_score(
        true_labels,
        true_predictions
    )

    f1 = f1_score(
        true_labels,
        true_predictions
    )

    return {
        "precision": precision,
        "recall": recall,
        "f1": f1
    }

In [40]:
import torch
import transformers

print("Transformers version:", transformers.__version__)
print("PyTorch version:", torch.__version__)
print("CUDA available:", torch.cuda.is_available())

if torch.cuda.is_available():
    print("GPU:", torch.cuda.get_device_name(0))
else:
    print("Training device: CPU")

Transformers version: 5.16.1
PyTorch version: 2.14.0+cpu
CUDA available: False
Training device: CPU


In [41]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="../models/clinical_ner_smoke_test",

    num_train_epochs=1,

    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,

    learning_rate=5e-5,

    logging_steps=5,

    save_strategy="no",

    report_to="none"
)

In [42]:
%pip install accelerate>=1.1.0

Note: you may need to restart the kernel to use updated packages.


In [43]:
smoke_train_dataset = train_dataset.select(range(10))
smoke_val_dataset = val_dataset.select(range(5))

print("Smoke train:", len(smoke_train_dataset))
print("Smoke validation:", len(smoke_val_dataset))

Smoke train: 10
Smoke validation: 5


In [44]:
from transformers import Trainer

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=smoke_train_dataset,
    eval_dataset=smoke_val_dataset,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)

In [45]:
trainer.train()

c:\Users\gangu\healthcare-clinical-nlp\.venv\Lib\site-packages\torch\utils\data\dataloader.py:759: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  super().__init__(loader)


Step,Training Loss
5,3.632330


TrainOutput(global_step=5, training_loss=3.632329559326172, metrics={'train_runtime': 26.1647, 'train_samples_per_second': 0.382, 'train_steps_per_second': 0.191, 'total_flos': 2589321744360.0, 'train_loss': 3.632329559326172, 'epoch': 1.0})

In [47]:
import numpy as np

In [48]:
smoke_results = trainer.evaluate()

print(smoke_results)

c:\Users\gangu\healthcare-clinical-nlp\.venv\Lib\site-packages\seqeval\metrics\v1.py:57: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Training Loss,Validation Loss,Step,Precision,Recall,F1
3.632330,2.877145,5,0.000000,0.000000,0.000000


{'eval_loss': 2.8771445751190186, 'eval_precision': 0.0, 'eval_recall': 0.0, 'eval_f1': 0.0}


In [49]:
import inspect
from transformers import TrainingArguments

print(inspect.signature(TrainingArguments.__init__))

(self, output_dir: str | None = None, per_device_train_batch_size: int = 8, num_train_epochs: float = 3.0, max_steps: int = -1, learning_rate: float = 5e-05, lr_scheduler_type: transformers.trainer_utils.SchedulerType | str = 'linear', lr_scheduler_kwargs: dict | str | None = None, warmup_steps: float = 0, optim: transformers.training_args.OptimizerNames | str = 'adamw_torch_fused', optim_args: str | None = None, weight_decay: float = 0.0, adam_beta1: float = 0.9, adam_beta2: float = 0.999, adam_epsilon: float = 1e-08, optim_target_modules: None | str | list[str] = None, gradient_accumulation_steps: int = 1, average_tokens_across_devices: bool = True, max_grad_norm: float = 1.0, label_smoothing_factor: float = 0.0, bf16: bool = False, fp16: bool = False, bf16_full_eval: bool = False, fp16_full_eval: bool = False, tf32: bool | None = None, gradient_checkpointing: bool = False, gradient_checkpointing_kwargs: dict[str, typing.Any] | str | None = None, torch_compile: bool = False, torch_co

In [ ]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="../models/clinical_ner",

    num_train_epochs=3,

    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,

    gradient_accumulation_steps=2,

    learning_rate=5e-5,
    weight_decay=0.01,

    eval_strategy="epoch",
    save_strategy="epoch",

    load_best_model_at_end=True,
    metric_for_best_model="f1",
    greater_is_better=True,

    logging_steps=20,

    save_total_limit=2,

    report_to="none",

    dataloader_pin_memory=False,

    seed=42
)

In [51]:
from transformers import Trainer

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)

In [53]:
from transformers import AutoModelForTokenClassification

model = AutoModelForTokenClassification.from_pretrained(
    model_name,
    num_labels=len(all_labels),
    id2label=id2label,
    label2id=label2id
)

print("Number of labels:", model.config.num_labels)

Loading weights: 100%|██████████| 197/197 [00:00<00:00, 22025.85it/s]
[transformers] BertForTokenClassification LOAD REPORT from: emilyalsentzer/Bio_ClinicalBERT
Key                                        | Status     | 
-------------------------------------------+------------+-
cls.predictions.bias                       | UNEXPECTED | 
cls.seq_relationship.bias                  | UNEXPECTED | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED | 
bert.pooler.dense.weight                   | UNEXPECTED | 
cls.predictions.decoder.weight             | UNEXPECTED | 
cls.seq_relationship.weight                | UNEXPECTED | 
bert.pooler.dense.bias                     | UNEXPECTED | 
cls.predictions.transform.dense.weight     | UNEXPECTED | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED | 
cls.predictions.transform.dense.bias       | UNEXPECTED | 
classifier.weight                          | MISSING    | 
classifier.bias                            | MISSING    | 

Notes:
- UN

Number of labels: 82


In [54]:
from transformers import Trainer

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)

In [55]:
print("Training chunks:", len(trainer.train_dataset))
print("Validation chunks:", len(trainer.eval_dataset))
print("Epochs:", trainer.args.num_train_epochs)
print("Train batch size:", trainer.args.per_device_train_batch_size)
print("Gradient accumulation:", trainer.args.gradient_accumulation_steps)

Training chunks: 286
Validation chunks: 64
Epochs: 3
Train batch size: 2
Gradient accumulation: 2


In [56]:
trainer.train()

Epoch,Training Loss,Validation Loss,Precision,Recall,F1
1,3.448005,1.388833,0.324588,0.297890,0.310667
2,2.095623,0.975401,0.475612,0.542297,0.506770
3,1.730202,0.916346,0.494667,0.573194,0.531043


Writing model shards: 100%|██████████| 1/1 [00:00<00:00,  1.15it/s]


TrainOutput(global_step=216, training_loss=2.6875117973045066, metrics={'train_runtime': 71004.7241, 'train_samples_per_second': 0.012, 'train_steps_per_second': 0.003, 'total_flos': 210955668584880.0, 'train_loss': 2.6875117973045066, 'epoch': 3.0})

In [58]:
test_results = trainer.evaluate(eval_dataset=test_dataset, metric_key_prefix="test")

print(test_results)

Training Loss,Validation Loss,Epoch,Precision,Recall,F1
1.730202,0.895174,3,0.504885,0.574911,0.537627


{'test_loss': 0.8951736092567444, 'test_precision': 0.5048846393681148, 'test_recall': 0.5749112426035503, 'test_f1': 0.537627268702966}


In [59]:
test_output = trainer.predict(test_dataset)
print("Predictions shape:", test_output.predictions.shape)
print("Labels shape:", test_output.label_ids.shape)

Predictions shape: (59, 512, 82)
Labels shape: (59, 512)


In [60]:
predicted_ids = np.argmax(
    test_output.predictions,
    axis=2
)

true_predictions = []
true_labels = []

for prediction, label in zip(
    predicted_ids,
    test_output.label_ids
):
    current_predictions = []
    current_labels = []

    for predicted_id, label_id in zip(
        prediction,
        label
    ):
        if label_id != -100:
            current_predictions.append(
                id2label[int(predicted_id)]
            )

            current_labels.append(
                id2label[int(label_id)]
            )

    true_predictions.append(current_predictions)
    true_labels.append(current_labels)

In [61]:
print("Number of sequences:", len(true_predictions))
print("First sequence predictions:", true_predictions[0][:20])
print("First sequence true labels:", true_labels[0][:20])

Number of sequences: 59
First sequence predictions: ['O', 'B-Age', 'I-Age', 'I-Age', 'I-Age', 'I-Age', 'B-Sex', 'O', 'B-Clinical_event', 'O', 'O', 'B-Nonbiological_location', 'O', 'B-Date', 'I-Date', 'O', 'O', 'B-Biological_structure', 'B-Sign_symptom', 'O']
First sequence true labels: ['O', 'B-Age', 'I-Age', 'I-Age', 'I-Age', 'I-Age', 'B-Sex', 'O', 'B-Clinical_event', 'O', 'O', 'B-Nonbiological_location', 'O', 'B-Date', 'I-Date', 'O', 'O', 'B-Sign_symptom', 'I-Sign_symptom', 'O']


In [62]:
from seqeval.metrics import classification_report

report = classification_report(
    true_labels,
    true_predictions,
    digits=3
)

print(report)

                        precision    recall  f1-score   support

              Activity      0.000     0.000     0.000         7
        Administration      0.000     0.000     0.000        33
                   Age      0.628     0.844     0.720        32
                  Area      0.000     0.000     0.000        15
  Biological_structure      0.562     0.687     0.619       486
        Clinical_event      0.692     0.736     0.714       110
                 Color      0.000     0.000     0.000         1
           Coreference      0.000     0.000     0.000        46
                  Date      0.381     0.701     0.493       107
  Detailed_description      0.309     0.390     0.345       523
  Diagnostic_procedure      0.628     0.717     0.670       771
      Disease_disorder      0.422     0.414     0.418       237
              Distance      0.400     0.071     0.121        28
                Dosage      0.319     0.578     0.411        64
              Duration      0.089     0

c:\Users\gangu\healthcare-clinical-nlp\.venv\Lib\site-packages\seqeval\metrics\v1.py:57: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [64]:
[name for name in dir() if "train" in name.lower()]

['Trainer',
 'TrainingArguments',
 'smoke_train_dataset',
 'tokenized_train_data',
 'train_chunked',
 'train_dataset',
 'train_ner_data',
 'trainer',
 'training_args']

In [65]:
from collections import Counter

train_entity_counts = Counter()

for document in train_ner_data:
    for label in document["labels"]:
        if label.startswith("B-"):
            entity_type = label[2:]
            train_entity_counts[entity_type] += 1

train_entity_counts

Counter({'Diagnostic_procedure': 3131,
         'Sign_symptom': 2346,
         'Lab_value': 1983,
         'Detailed_description': 1974,
         'Biological_structure': 1930,
         'Disease_disorder': 879,
         'Medication': 750,
         'Therapeutic_procedure': 671,
         'Date': 488,
         'Clinical_event': 445,
         'Dosage': 263,
         'Nonbiological_location': 246,
         'Severity': 242,
         'Coreference': 238,
         'Duration': 204,
         'History': 181,
         'Age': 142,
         'Administration': 134,
         'Sex': 133,
         'Distance': 71,
         'Activity': 60,
         'Family_history': 51,
         'Shape': 49,
         'Color': 43,
         'Time': 42,
         'Frequency': 40,
         'Personal_background': 38,
         'Texture': 35,
         'Subject': 35,
         'Outcome': 30,
         'Volume': 24,
         'Qualitative_concept': 23,
         'Area': 21,
         'Other_event': 19,
         'Other_entity': 13,
        

In [66]:
import pandas as pd

report_dict = classification_report(
    true_labels,
    true_predictions,
    output_dict=True,
    zero_division=0
)

report_df = pd.DataFrame(report_dict).T

report_df["train_count"] = pd.Series(train_entity_counts)

analysis_df = report_df[
    ~report_df.index.isin(
        ["micro avg", "macro avg", "weighted avg"]
    )
].copy()

analysis_df = analysis_df[
    ["train_count", "support", "precision", "recall", "f1-score"]
]

analysis_df = analysis_df.sort_values(
    "train_count",
    ascending=False
)

print(analysis_df)

                        train_count  support  precision    recall  f1-score
Diagnostic_procedure         3131.0    771.0   0.628409  0.717250  0.669897
Sign_symptom                 2346.0    487.0   0.500697  0.737166  0.596346
Lab_value                    1983.0    466.0   0.561688  0.742489  0.639556
Detailed_description         1974.0    523.0   0.309091  0.390057  0.344886
Biological_structure         1930.0    486.0   0.562290  0.687243  0.618519
Disease_disorder              879.0    237.0   0.422414  0.413502  0.417910
Medication                    750.0    198.0   0.593640  0.848485  0.698545
Therapeutic_procedure         671.0    257.0   0.342105  0.202335  0.254279
Date                          488.0    107.0   0.380711  0.700935  0.493421
Clinical_event                445.0    110.0   0.692308  0.736364  0.713656
Dosage                        263.0     64.0   0.318966  0.578125  0.411111
Nonbiological_location        246.0     66.0   0.596774  0.560606  0.578125
Severity    

In [67]:
spearman_corr = analysis_df[
    ["train_count", "f1-score"]
].corr(method="spearman")

print(spearman_corr)

             train_count  f1-score
train_count     1.000000  0.780147
f1-score        0.780147  1.000000


In [68]:
def tokenize_test_with_metadata(item):
    tokens = item["tokens"]
    labels = item["labels"]

    tokenized = tokenizer(
        tokens,
        is_split_into_words=True,
        truncation=True,
        max_length=512,
        stride=128,
        return_overflowing_tokens=True
    )

    chunks = []

    for chunk_index in range(len(tokenized["input_ids"])):

        word_ids = tokenized.word_ids(
            batch_index=chunk_index
        )

        chunks.append({
            "document_id": item["document_id"],
            "chunk_index": chunk_index,
            "input_ids": tokenized["input_ids"][chunk_index],
            "attention_mask": tokenized["attention_mask"][chunk_index],
            "word_ids": word_ids
        })

    return chunks

In [69]:
sample_test_item = test_ner_data[0]

sample_chunks = tokenize_test_with_metadata(
    sample_test_item
)

print("Document ID:", sample_test_item["document_id"])
print("Original words:", len(sample_test_item["tokens"]))
print("Number of chunks:", len(sample_chunks))

for chunk in sample_chunks:
    valid_word_ids = [
        word_id
        for word_id in chunk["word_ids"]
        if word_id is not None
    ]

    print(
        f"Chunk {chunk['chunk_index']}:",
        f"length={len(chunk['input_ids'])},",
        f"word range={min(valid_word_ids)}-{max(valid_word_ids)}"
    )

Document ID: 18416479
Original words: 592
Number of chunks: 2
Chunk 0: length=512, word range=0-364
Chunk 1: length=450, word range=285-590


In [70]:
print("Last 10 original tokens:")
print(list(enumerate(
    sample_test_item["tokens"][-10:],
    start=len(sample_test_item["tokens"]) - 10
)))

covered_word_ids = set()

for chunk in sample_chunks:
    covered_word_ids.update(
        word_id
        for word_id in chunk["word_ids"]
        if word_id is not None
    )

missing_word_ids = set(
    range(len(sample_test_item["tokens"]))
) - covered_word_ids

print("\nMissing word IDs:", missing_word_ids)

Last 10 original tokens:
[(582, 'up'), (583, 'of'), (584, 'esophageal'), (585, 'varices'), (586, 'without'), (587, 'prophylactic'), (588, 'endoscopic'), (589, 'therapy'), (590, '.'), (591, '\n')]

Missing word IDs: {261, 395, 144, 272, 532, 408, 25, 416, 161, 290, 37, 551, 302, 563, 55, 444, 322, 67, 591, 208, 472, 93, 482, 227, 492, 240, 116, 250}


In [71]:
for word_id in sorted(missing_word_ids):
    token = sample_test_item["tokens"][word_id]

    print(
        word_id,
        repr(token),
        "whitespace =", token.isspace()
    )

25 '\n' whitespace = True
37 '\n' whitespace = True
55 '\n' whitespace = True
67 '\n' whitespace = True
93 '\n' whitespace = True
116 '\n' whitespace = True
144 '\n' whitespace = True
161 '\n' whitespace = True
208 '\n' whitespace = True
227 '\n' whitespace = True
240 '\n' whitespace = True
250 '\n' whitespace = True
261 '\n' whitespace = True
272 '\n' whitespace = True
290 '\n' whitespace = True
302 '\n' whitespace = True
322 '\n' whitespace = True
395 '\n' whitespace = True
408 '\n' whitespace = True
416 '\n' whitespace = True
444 '\n' whitespace = True
472 '\n' whitespace = True
482 '\n' whitespace = True
492 '\n' whitespace = True
532 '\n' whitespace = True
551 '\n' whitespace = True
563 '\n' whitespace = True
591 '\n' whitespace = True


In [72]:
import torch

model.eval()

sample_predictions = []

for chunk in sample_chunks:

    input_ids = torch.tensor(
        [chunk["input_ids"]]
    ).to(model.device)

    attention_mask = torch.tensor(
        [chunk["attention_mask"]]
    ).to(model.device)

    with torch.no_grad():
        outputs = model(
            input_ids=input_ids,
            attention_mask=attention_mask
        )

    logits = outputs.logits[0].cpu()

    sample_predictions.append({
        "document_id": chunk["document_id"],
        "chunk_index": chunk["chunk_index"],
        "word_ids": chunk["word_ids"],
        "logits": logits
    })

print("Chunks predicted:", len(sample_predictions))

for prediction in sample_predictions:
    print(
        "Chunk:",
        prediction["chunk_index"],
        "Logits shape:",
        prediction["logits"].shape
    )

Chunks predicted: 2
Chunk: 0 Logits shape: torch.Size([512, 82])
Chunk: 1 Logits shape: torch.Size([450, 82])


In [73]:
from collections import defaultdict
import torch

word_logits = defaultdict(list)

for prediction in sample_predictions:

    word_ids = prediction["word_ids"]
    logits = prediction["logits"]

    previous_word_id = None

    for position, word_id in enumerate(word_ids):

        if word_id is None:
            continue

        # Use only the first WordPiece of each original word
        if word_id == previous_word_id:
            continue

        word_logits[word_id].append(
            logits[position]
        )

        previous_word_id = word_id

In [74]:
single_count = 0
overlap_count = 0

for word_id, scores in word_logits.items():

    if len(scores) == 1:
        single_count += 1

    elif len(scores) > 1:
        overlap_count += 1

print("Unique tokenized words:", len(word_logits))
print("Words appearing once:", single_count)
print("Words appearing in overlap:", overlap_count)

Unique tokenized words: 564
Words appearing once: 487
Words appearing in overlap: 77


In [75]:
final_predicted_labels = {}

for word_id, scores in word_logits.items():

    stacked_scores = torch.stack(scores)

    average_scores = stacked_scores.mean(dim=0)

    predicted_id = torch.argmax(average_scores).item()

    final_predicted_labels[word_id] = id2label[predicted_id]

In [76]:
document_predictions = []
document_true_labels = []

for word_id in sorted(final_predicted_labels.keys()):

    document_predictions.append(
        final_predicted_labels[word_id]
    )

    document_true_labels.append(
        sample_test_item["labels"][word_id]
    )

print("Predictions:", len(document_predictions))
print("True labels:", len(document_true_labels))

print("\nFirst 20 predictions:")
print(document_predictions[:20])

print("\nFirst 20 true labels:")
print(document_true_labels[:20])

Predictions: 564
True labels: 564

First 20 predictions:
['O', 'B-Age', 'I-Age', 'I-Age', 'I-Age', 'I-Age', 'B-Sex', 'O', 'B-Clinical_event', 'O', 'O', 'B-Nonbiological_location', 'O', 'B-Date', 'I-Date', 'O', 'O', 'B-Biological_structure', 'B-Sign_symptom', 'O']

First 20 true labels:
['O', 'B-Age', 'I-Age', 'I-Age', 'I-Age', 'I-Age', 'B-Sex', 'O', 'B-Clinical_event', 'O', 'O', 'B-Nonbiological_location', 'O', 'B-Date', 'I-Date', 'O', 'O', 'B-Sign_symptom', 'I-Sign_symptom', 'O']


In [77]:
def predict_document_overlap_aware(item):

    chunks = tokenize_test_with_metadata(item)

    word_logits = defaultdict(list)

    model.eval()

    for chunk in chunks:

        input_ids = torch.tensor(
            [chunk["input_ids"]]
        ).to(model.device)

        attention_mask = torch.tensor(
            [chunk["attention_mask"]]
        ).to(model.device)

        with torch.no_grad():
            outputs = model(
                input_ids=input_ids,
                attention_mask=attention_mask
            )

        logits = outputs.logits[0].cpu()

        previous_word_id = None

        for position, word_id in enumerate(chunk["word_ids"]):

            if word_id is None:
                continue

            if word_id == previous_word_id:
                continue

            word_logits[word_id].append(
                logits[position]
            )

            previous_word_id = word_id

    predicted_labels = []
    true_labels_document = []

    for word_id in sorted(word_logits.keys()):

        stacked_scores = torch.stack(
            word_logits[word_id]
        )

        average_scores = stacked_scores.mean(
            dim=0
        )

        predicted_id = torch.argmax(
            average_scores
        ).item()

        predicted_labels.append(
            id2label[predicted_id]
        )

        true_labels_document.append(
            item["labels"][word_id]
        )

    return predicted_labels, true_labels_document

In [78]:
document_predictions_all = []
document_true_labels_all = []

for i, item in enumerate(test_ner_data):

    predictions, labels = predict_document_overlap_aware(
        item
    )

    document_predictions_all.append(
        predictions
    )

    document_true_labels_all.append(
        labels
    )

    print(
        f"Document {i + 1}/{len(test_ner_data)}:",
        len(predictions),
        "words"
    )

Document 1/30: 564 words
Document 2/30: 310 words
Document 3/30: 770 words
Document 4/30: 516 words
Document 5/30: 754 words
Document 6/30: 410 words
Document 7/30: 691 words
Document 8/30: 522 words
Document 9/30: 301 words
Document 10/30: 663 words
Document 11/30: 289 words
Document 12/30: 383 words
Document 13/30: 536 words
Document 14/30: 511 words
Document 15/30: 516 words
Document 16/30: 350 words
Document 17/30: 650 words
Document 18/30: 299 words
Document 19/30: 471 words
Document 20/30: 375 words
Document 21/30: 527 words
Document 22/30: 516 words
Document 23/30: 579 words
Document 24/30: 295 words
Document 25/30: 426 words
Document 26/30: 612 words
Document 27/30: 522 words
Document 28/30: 762 words
Document 29/30: 350 words
Document 30/30: 545 words


In [79]:
from seqeval.metrics import (
    precision_score,
    recall_score,
    f1_score
)

document_precision = precision_score(
    document_true_labels_all,
    document_predictions_all
)

document_recall = recall_score(
    document_true_labels_all,
    document_predictions_all
)

document_f1 = f1_score(
    document_true_labels_all,
    document_predictions_all
)

print("Overlap-aware document-level results")
print("------------------------------------")
print(f"Precision: {document_precision:.4f}")
print(f"Recall:    {document_recall:.4f}")
print(f"F1:        {document_f1:.4f}")

Overlap-aware document-level results
------------------------------------
Precision: 0.5083
Recall:    0.5770
F1:        0.5405


# Final Model Development and Evaluation

## Project Objective

The objective of this experiment was to fine-tune BioClinicalBERT for
fine-grained clinical Named Entity Recognition (NER).

The model identifies clinical concepts such as diseases, symptoms,
medications, procedures, laboratory values, anatomical structures,
dosages, dates, and other clinical entities from clinical text.

The dataset contains 200 annotated clinical documents with 41 clinical
entity categories.

BIO encoding was used to represent entity boundaries:

- B = Beginning of an entity
- I = Inside an entity
- O = Outside an entity

A total of 82 BIO labels were observed in the training data.

---

## Dataset Split

Documents were split at the document level to prevent information leakage.

| Dataset | Documents | Percentage |
|---|---:|---:|
| Training | 140 | 70% |
| Validation | 30 | 15% |
| Test | 30 | 15% |
| Total | 200 | 100% |

Random seed: 42

The held-out test set was not used during model development or
hyperparameter selection.

---

## Base Model

Pretrained model:

`emilyalsentzer/Bio_ClinicalBERT`

BioClinicalBERT was selected because it provides contextual language
representations pretrained on clinical text.

A token-classification head was added for the 82 observed BIO labels and
fine-tuned using the annotated training data.

---

## Tokenization and BIO Label Alignment

BioClinicalBERT uses WordPiece tokenization, which can divide one original
word into multiple subword tokens.

For example:

Original word:

`palpitations`

WordPiece representation:

`p`, `##al`, `##pit`, `##ations`

To prevent one entity label from being counted multiple times, the original
BIO label was assigned only to the first WordPiece.

Continuation WordPieces and special tokens such as `[CLS]` and `[SEP]`
were assigned a label value of `-100`, causing them to be ignored during
loss calculation and evaluation.

---

## Long-Document Processing

Many clinical documents exceeded BERT's maximum sequence length of
512 tokens.

Instead of truncating these documents, overlapping windows were used:

- Maximum sequence length: 512
- Stride: 128

This allowed information near chunk boundaries to appear in more than
one window and prevented large portions of long clinical documents from
being discarded.

Training documents were transformed into 286 chunks and validation
documents into 64 chunks.

---

# Model Development

## Selected Training Configuration

The final model configuration was:

| Parameter | Value |
|---|---|
| Base model | BioClinicalBERT |
| Learning rate | 5e-5 |
| Epochs | 5 |
| Training batch size | 2 |
| Gradient accumulation | 2 |
| Effective batch size | 4 |
| Weight decay | 0.01 |
| Maximum sequence length | 512 |
| Stride | 128 |
| Random seed | 42 |
| Model selection metric | Validation entity-level F1 |

The best validation checkpoint occurred at Epoch 4.

---

## Validation Performance

The selected model achieved the following overlap-aware document-level
validation performance:

| Metric | Score |
|---|---:|
| Precision | 0.5505 |
| Recall | 0.6176 |
| F1 | 0.5821 |

These metrics are entity-level NER metrics rather than ordinary
classification accuracy.

---

# Controlled Experiments

Several controlled experiments were conducted before final model selection.

## Five Epoch Training

Increasing training from three to five epochs improved validation
performance.

The best validation performance occurred at Epoch 4.

Validation F1:

`0.5821`

Performance stopped improving after this point, suggesting that additional
epochs were unlikely to provide meaningful improvement under the same
configuration.

---

## Lower Learning Rate

A learning rate of `3e-5` was evaluated while keeping the remaining
training configuration similar.

The resulting validation F1 was approximately:

`0.540`

This did not outperform the `5e-5` configuration and was therefore not
selected.

---

## Class-Weighted Loss

The training data contained substantial class imbalance. Some entity
categories occurred thousands of times while others had very few examples.

A class-weighted cross-entropy experiment was therefore conducted to test
whether giving additional importance to less frequent labels would improve
performance.

Results:

| Metric | Weighted Model |
|---|---:|
| Precision | 0.5139 |
| Recall | 0.6190 |
| F1 | 0.5616 |

The weighted model slightly increased recall but reduced precision and
overall F1.

Therefore, class weighting was rejected and the standard cross-entropy
model was retained.

---

# Error Analysis

Error analysis was performed to understand the model's limitations rather
than relying only on the aggregate F1 score.

Among analyzed incorrect token-level predictions:

| Error Type | Percentage |
|---|---:|
| Wrong entity type | 49.32% |
| Entity ↔ O error | 41.19% |
| BIO boundary-only error | 9.49% |

The relatively small proportion of boundary-only errors suggested that
incorrect entity boundaries were not the dominant problem.

The largest source of error was semantic confusion between clinical entity
categories.

Examples of frequently confused categories included:

- Disease_disorder → Sign_symptom
- Diagnostic_procedure → Detailed_description
- Detailed_description → Biological_structure
- Sign_symptom → Disease_disorder
- Family_history → History

This indicates that future improvement should focus on semantic
discrimination between closely related clinical concepts rather than only
BIO boundary prediction.

---

# Overlap-Aware Document Reconstruction

Because long documents were processed using overlapping chunks, the same
original word could appear in multiple chunks.

During document-level evaluation:

1. Predictions were mapped from WordPiece tokens back to original words.
2. Only the first WordPiece of each original word was evaluated.
3. Logits for words appearing in multiple overlapping chunks were averaged.
4. The label with the highest averaged logit was selected.
5. Original document-level prediction sequences were reconstructed.
6. Entity-level precision, recall, and F1 were calculated using seqeval.

Tokenizer-dropped whitespace/newline tokens that produced no WordPiece were
excluded from reconstruction.

This is an evaluation limitation because some dropped tokens contained BIO
labels and their removal can affect reconstructed BIO sequences.

---

# Final Held-Out Test Evaluation

After model development and selection were completed, the selected model
was evaluated on the untouched 30-document test set.

No model-selection decisions were made using the test results.

## Final Test Results

| Metric | Score |
|---|---:|
| Precision | **0.5805** |
| Recall | **0.6500** |
| Entity-level Micro F1 | **0.6133** |

The selected BioClinicalBERT model therefore achieved:

**61.33% entity-level micro F1 on the held-out test set using
overlap-aware document-level reconstruction.**

The test F1 was slightly higher than the validation F1. Given the relatively
small validation and test sets, variation between the two document samples
is expected.

---

# Final Model Selection

The final selected approach was:

BioClinicalBERT  
→ 82 observed BIO labels  
→ WordPiece-aware label alignment  
→ 512-token overlapping windows  
→ 128-token stride  
→ 5-epoch training schedule  
→ best validation checkpoint  
→ overlap-aware document reconstruction

Final held-out performance:

**Precision: 58.05%**  
**Recall: 65.00%**  
**Entity-level Micro F1: 61.33%**

The model should be considered an experimental clinical NLP model rather
than a production clinical decision-support system.

---

# Key Findings

1. BioClinicalBERT successfully learned fine-grained clinical entity
   extraction across a large entity taxonomy.

2. Long clinical documents required explicit handling because many exceeded
   BERT's 512-token context limit.

3. Increasing training duration improved performance until validation
   performance plateaued.

4. Lowering the learning rate did not improve the selected model.

5. Class-weighted loss increased recall slightly but reduced precision and
   overall F1.

6. Error analysis showed that semantic entity-type confusion was a much
   larger problem than BIO boundary errors.

7. The final selected model generalized to the held-out test set with an
   entity-level micro F1 of 0.6133.

---

# Limitations and Future Work

The model still makes meaningful errors and should not be considered
production-ready for clinical decision-making.

Potential future improvements include:

- reviewing annotation consistency for frequently confused entity types
- increasing examples for underrepresented entity categories
- investigating clinically meaningful taxonomy simplification
- comparing stronger clinical language-model encoders
- improving long-document contextual modeling
- exploring span-based NER or structured decoding approaches
- evaluating performance on additional external clinical datasets

The next stage of this project focuses on integrating the NER model into a
Clinical Document Review Assistant that converts extracted clinical
entities into structured, interpretable information.